icechunk/src/store.rs

```diff
@@ -428,31 +426,100 @@ impl Store {
         &self,
         prefix: &str,
     ) -> StoreResult<impl Stream<Item = StoreResult<ListDirItem>> + Send> {
-        // TODO: this is inefficient because it filters based on the prefix, instead of only
-        // generating items that could potentially match
-        // FIXME: this is not lazy, it goes through every chunk. This should be implemented using
-        // metadata only, and ignore the chunks, but we should decide on that based on Zarr3 spec
-        // evolution
-
-        let idx: usize = if prefix == "/" { 0 } else { prefix.len() };
-
-        let parents: HashSet<_> = self
-            .list_prefix(prefix)
-            .await?
-            .map_ok(move |s| {
-                // If the prefix is "/", get rid of it. This can happen when prefix is missing
-                // the trailing slash (as it does in zarr-python impl)
-                let rem = &s[idx..].trim_start_matches('/');
-                match rem.split_once('/') {
-                    Some((prefix, _)) => ListDirItem::Prefix(prefix.to_string()),
-                    None => ListDirItem::Key(rem.to_string()),
+        let session = Arc::clone(&self.session).read_owned().await;
+        let prefix = prefix.trim_end_matches("/");
+        let absolute_prefix =
+            if !prefix.starts_with("/") { &format!("/{}", prefix) } else { prefix };
+
+        let path = Path::try_from(absolute_prefix)?;
+        let results = match session.get_node(&path).await {
+            Ok(NodeSnapshot {
+                id: node_id,
+                path: node_path,
+                node_data: NodeData::Array(..),
+                ..
+            }) => {
+                // if this is an array we know what to yield
+                let mut res = vec![ListDirItem::Key("zarr.json".to_string())];
+                // The `c` prefix will exist when an array was created and a chunk was written
+                // either (1) in the changeset or (2) as part of a previous snapshot.
+                // It will not exist if (3) the array is recorded as deleted in the change_set.
+                if session.array_has_modified_chunks(node_path, node_id).await? {
```

Comment:
maybe we are working too hard for something that doesn't really matter? What happens if we always include the `/c`? I don't imagine anything will break.